src/agent/advertising_proxy.cpp

```diff
@@ -0,0 +1,338 @@
+ */
+
+/**
+ * @file
+ *   The file implements the Advertising Proxy.
+ */
+
+#include "agent/advertising_proxy.hpp"
+
+#include <string>
+
+#include <assert.h>
+
+#include "common/code_utils.hpp"
+#include "common/logging.hpp"
+
+#if OTBR_ENABLE_ADVERTISING_PROXY
+
+namespace otbr {
+
+static otError OtbrErrorToOtError(otbrError aError)
+{
+    otError error;
+
+    switch (aError)
+    {
+    case OTBR_ERROR_NONE:
+        error = OT_ERROR_NONE;
+        break;
+
+    case OTBR_ERROR_NOT_FOUND:
+        error = OT_ERROR_NOT_FOUND;
+        break;
+
+    case OTBR_ERROR_PARSE:
+        error = OT_ERROR_PARSE;
+        break;
+
+    case OTBR_ERROR_NOT_IMPLEMENTED:
+        error = OT_ERROR_NOT_IMPLEMENTED;
+        break;
+
+    case OTBR_ERROR_INVALID_ARGS:
+        error = OT_ERROR_INVALID_ARGS;
+        break;
+
+    case OTBR_ERROR_DUPLICATED:
+        error = OT_ERROR_DUPLICATED;
+        break;
+
+    default:
+        error = OT_ERROR_FAILED;
+        break;
+    }
+
+    return error;
+}
+
+static otbrError SplitFullServiceName(const std::string &aFullName,
+                                      std::string &      aInstanceName,
+                                      std::string &      aType,
+                                      std::string &      aDomain)
+{
+    otbrError error = OTBR_ERROR_INVALID_ARGS;
+    size_t    dotPos[3];
+
+    dotPos[0] = aFullName.find_first_of('.');
+    VerifyOrExit(dotPos[0] != std::string::npos);
+
+    dotPos[1] = aFullName.find_first_of('.', dotPos[0] + 1);
+    VerifyOrExit(dotPos[1] != std::string::npos);
+
+    dotPos[2] = aFullName.find_first_of('.', dotPos[1] + 1);
+    VerifyOrExit(dotPos[2] != std::string::npos);
+
+    error         = OTBR_ERROR_NONE;
+    aInstanceName = aFullName.substr(0, dotPos[0]);
+    aType         = aFullName.substr(dotPos[0] + 1, dotPos[2] - dotPos[0] - 1);
+    aDomain       = aFullName.substr(dotPos[2] + 1, aFullName.size() - dotPos[2] - 1);
+
+exit:
+    return error;
+}
+
+static otbrError SplitFullHostName(const std::string &aFullName, std::string &aInstanceName, std::string &aDomain)
+{
+    otbrError error = OTBR_ERROR_INVALID_ARGS;
+    size_t    dotPos;
+
+    dotPos = aFullName.find_first_of('.');
+    VerifyOrExit(dotPos != std::string::npos);
+
+    error         = OTBR_ERROR_NONE;
+    aInstanceName = aFullName.substr(0, dotPos);
+    aDomain       = aFullName.substr(dotPos + 1, aFullName.size() - dotPos - 1);
+
+exit:
+    return error;
+}
+
+AdvertisingProxy::AdvertisingProxy(Mdns::Publisher *aPublisher)
+    : mInstance(nullptr)
+    , mPublisher(aPublisher)
+{
+}
+
+otbrError AdvertisingProxy::Start(otInstance *aInstance)
+{
+    mInstance = aInstance;
+
+    otSrpServerSetServiceUpdateHandler(mInstance, AdvertisingHandler, this);
+
+    mPublisher->SetPublishServiceHandler(PublishServiceHandler, this);
+    mPublisher->SetPublishHostHandler(PublishHostHandler, this);
+
+    otbrLog(OTBR_LOG_INFO, "[adproxy] Advertising Proxy Started");
```

Comment:
nit: seems redundant

```suggestion
    otbrLog(OTBR_LOG_INFO, "[adproxy] Started");
```